keepercommander/importer/imp_exp.py

```diff
@@ -678,211 +1064,269 @@ def prepare_folder_add(params, folders, records):
     return folder_add
 
 
-def tokenize_partial_import_record(record):   # type: (ImportRecord) -> [str]
+def parepare_record_audit(params, uids, enterprise_public_key):  # type: (KeeperParams, Iterator[str], any) -> Iterator[record_proto.RecordAddAuditData]
+    for uid in uids:
+        if uid in params.record_cache:
+            keeper_record = params.record_cache[uid]
+            import_record = convert_keeper_record(keeper_record)
+            if import_record:
+                audit_data = {
+                    'title': import_record.title,
+                    'record_type': import_record.type
+                }
+                if import_record.login_url:
+                    audit_data['url'] = import_record.login_url
+                record_audit_rq = record_proto.RecordAddAuditData()
+                record_audit_rq.record_uid = utils.base64_url_decode(uid)
+                record_audit_rq.revision = 0
+                record_audit_rq.data = crypto.encrypt_ec(json.dumps(audit_data).encode('utf-8'), enterprise_public_key)
+                yield record_audit_rq
+
+
+RECORD_FIELD_TYPES = {'login', 'password', 'url'}
+IGNORABLE_FIELD_TYPES = {'text', 'fileRef', 'cardRef', 'oneTimeCode'}
+
+
+def value_to_token(value): # type: (any) -> str
+    if not value:
+        return ''
+    if type(value) == str:
+        return value
+    if type(value) == list:
+        return ','.join((value_to_token(x) for x in value))
+    if type(value) == dict:
+        pairs = [x for x in value.items()]
+        pairs.sort(key=lambda x: x[0])
+        return ';'.join(f'{k}={value_to_token(v)}' for k, v in pairs)
+    return str(value)
+
+
+def tokenize_record_key(record):   # type: (ImportRecord) -> Iterator[str]
     """
     Turn a record-to-import into an iterable of str's for hashing.  This is really about import --update.
 
     Examine just the relevant parts of the record.
     """
-    yield record.title or ''
-    yield record.login or ''
-    yield record.login_url or ''
-
-
-def tokenize_partial_preexisting_record(record):   # type: (Record) -> [str]
-    """
-    Turn a preexisting record into an iterable of str's for hashing.
-
-    Examine just the relevant parts of the record.
-
-    For now at least, this is the same as tokenize_partial_import_record().
-    """
-    return tokenize_partial_import_record(record)
+    record_type = record.type or ""
+    yield f'$type:{record_type}'
```

Comment:
What is the meaning of all these dollar signs?  This is more of a cross-training thing. It probably merits a comment though.

Reply:
`My Vault> rti -lr login                                                                                                                                                                                                            
  Record Type ID  Content
----------------  ----------------------------------
               1  {
                    "$id": "login",
                    "categories": [
                      "login"
                    ],
                    "description": "Login template",
                    "fields": [
                      {
                        "$ref": "login"
                      },
                      {
                        "$ref": "password"
                      },
                      {
                        "$ref": "url"
                      },
                      {
                        "$ref": "fileRef"
                      },
                      {
                        "$ref": "oneTimeCode"
                      }
                    ]
                  }
`

Reply:
For some reason, it was decided to mark types for "record V3" with $